include/private/neubot/dash_impl.hpp

```diff
@@ -0,0 +1,547 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+#ifndef PRIVATE_NEUBOT_DASH_IMPL_HPP
+#define PRIVATE_NEUBOT_DASH_IMPL_HPP
+
+/*
+ * Here we implement the measurement methodology described in "Measuring
+ * DASH Streaming Performance from the End Users Perspective using Neubot",
+ * by S. Basso, A. Servetti, E. Masala, J.C. De Martin.
+ *
+ * See: https://nexa.polito.it/publications/basso2014measuring
+ *
+ * Differences from the methodology outlined in the paper:
+ *
+ * 1. By default, we don't scale down significantly the speed after we
+ *    sense some queue in the network. That was meant to avoid clogging
+ *    the downlink for Neubot (a background tool), but for OONI we do
+ *    actually plan to run the test in the foreground. To re-enable the
+ *    old behavior, set the `fast_scale_down` option to non-zero.
+ *
+ * 2. By default, we don't use the vector of fixed rates but rather we
+ *    use as rate the speed computed in the previous set. This should
+ *    yield a more accurate streaming speed estimate. This different
+ *    behavior was suggested to me by Antonio Servetti sometime around
+ *    May '17. To re-enable the old behavior, you need to set the
+ *    `use_fixed_rates` option to non-zero.
+ *
+ * 3. As pointed out also by a comment in the code, our measurement of
+ *    the download speed starts when we send the request rather when we
+ *    receive the first byte of the response. I think this is not bad,
+ *    since it actually captures better the time to fetch the chunk that
+ *    one should play, as opposed to accurately estimating the speed.
+ *
+ * 4. We leave to zero the `delta_user_time` and `delta_sys_time`. We
+ *    may change this in the future, when we understand how to get them
+ *    portably. They were not so important for us anyway.
+ *
+ * 5. As pointed out also by a comment in the code, we don't consider
+ *    the overhead of protocols when computing the speed. As for point
+ *    three above, this allows to write a simpler implementation than
+ *    the one in Neubot and, as a byproduct, provides us with information
+ *    closer to know how much time it takes to get a playable chunk
+ *    than just measuring the raw speed (I think, when I wrote the test
+ *    for Neubot, I had too much bias towards measuring speed).
```

Comment:
TODO: also mention here (as documented in the code) that, when we're not using fixed rates, we start with an initial estimate of 3,000 kbit/s of download speed.